### ISSUE TITLE
refreshing quickly with aws-lambda streaming response triggers file download

### ISSUE THREAD

**Issue by ap0nia** [jill64/sveltekit-adapter-aws#540]
# Summary
Refreshing the `/sverdle` route on a default sverdle project created by `pnpm create svelte` can trigger a download to occur for some reason. Here's an [example deployment](https://d1h8gum5ra2no2.cloudfront.net/sverdle) that I experience this issue with. An empty file is downloaded if the page is refreshed quickly.

# Steps to reproduce
1. Create a new sverdle project with the `create svelte` command.
```sh
pnpm create svelte
```
2. Install the adapter.
```sh
pnpm i -D @jill64/sveltekit-adapter-aws
```
3. Update the svelte config.
```ts
// svelte.config.js
// @ts-check

import adapter from '@jill64/sveltekit-adapter-aws';
import { vitePreprocess } from '@sveltejs/vite-plugin-svelte';

/**
 * @type{import('@sveltejs/kit').Config}
 */
const config = {
	preprocess: [vitePreprocess()],
	kit: {
		adapter: adapter({
			name: 'sverdle',
			deploy: true,
			architecture: 'lambda-s3'
		})
	}
};

export default config;
```
4. Build and deploy the app.
```sh
pnpm build
```

# Details
- It doesn't seem to occur when the Lambda is using the `buffered` response type.
- The `content-type` header is set to `application/octet-stream` and the SvelteKit `Server.respond` function seems to return a `304` response whenever this occurs.
- I'm not sure if this is reproducible, but I experienced this issue across different CDK deployments and also on different browsers while testing it out.
- I work with my own custom CDK setup, and I was interested in adopting the response streaming strategy demonstrated by this project, but I ran into this issue with both adapters for some reason.

**Good Response (page renders normally)**
<img width="678" alt="image" src="https://github.com/jill64/sveltekit-adapter-aws/assets/66761366/c76c4ff2-04c4-47b0-af00-f80764470917">

**Bad Response (triggers download)**
<img width="680" alt="image" src="https://github.com/jill64/sveltekit-adapter-aws/assets/66761366/ebe58e6b-248f-4890-8caa-5be831e041d5">

**Comment by anilmagzter** [comment]
Facing the same issue. How can we disable stream option and use only buffered for now?

**Comment by ap0nia** [comment]
I think I've figured out the root cause. AWS Lambda's `responseStream` doesn't seem to respect any metadata you set unless you actually write something. 

For example, here I indicate that the response should have the `text/html` content-type, even there aren't any contents.  But I get a response with content-type `application/octet` stream unless I actually perform any sort of write.
```ts
responseStream.setContentType('text/html')
// response.write('\n') /* Literally any write is needed?? */
responseStream.end()
```

I speculate that the SvelteKit server intends to send a `304` response with an empty body, indicating the browser to retrieve the page from cache, but this metadata is being ignored by the `responseStream` because the body is empty. 

In the posted images above, I keep getting a `200` response whenever the SvelteKit server indicated that it should've been a `304` response.

**Comment by jill64** [comment]
I agree with @ap0nia.
I have tried several patterns.

## OK

```js
responseStream.setContentType("text/html");
response.write("\n");
responseStream.end();
```
```js
responseStream.setContentType("text/html");
response.write("");
responseStream.end();
```

```js
responseStream.setContentType("text/html");
responseStream.write(new Uint8Array(0));
responseStream.end();
```

## NG

```js
responseStream.setContentType("text/html");
responseStream.end();
```

```js
responseStream.setContentType("text/html");
response.write(null);
responseStream.end();
```

```js
responseStream.setContentType("text/html");
responseStream.write(new Buffer());
responseStream.end();
```

Therefore, the implementation code is likely to be as follows

```diff
responseStream.setContentType("text/html");

if (!response.body) {
+   responseStream.write("");
    responseStream.end();
    return;
}

// ...
```

However, if I understand correctly, there are cases where `null` and empty string ("") should have different meanings in html response.
Do you think this solution would have side effects on the application?

**Comment by ap0nia** [comment]
If your proposed solution works (I'm not sure because I haven't tested anything), then I think it would at least be preferable to spontaneously triggering file downloads.

In the example of the `304` status code, the response is supposed to have no body and no `content-type` header whatsoever. When I was testing out writes on my end, I think writing even an empty string causes an empty HTML page to render. For the case of `304`, the most correct implementation might be something like this.

```ts
if (response.status === 304) {
  responseStream.setContentType('')
  responseStream.end()
}
```

Aside from the fact that this probably should've already worked with the existing code, it doesn't even seem possible to manipulate the content type and status unless non-null data is written.

I wonder if this is some upstream issue with the AWS Lambda streaming interface itself 🤔

**Comment by ap0nia** [comment]
Sure, let me know how the solutions turn out! 👍 

I don't think my personal implementation of the AWS adapter is as comprehensive, but it was just something I put together for quickly switching between deployment types. I think the main points were:

- [add a `stream` boolean flag to the adapter config](https://github.com/ap0nia/svelte.kit/blob/main/packages/cdk/src/index.ts#L69) this flag is read by both the CDK and adapter, and I just happened to assign it to the CDK package
- CDK -- [set the invoke mode on the function URL based on the flag value](https://github.com/ap0nia/svelte.kit/blob/main/packages/cdk/src/index.ts#L314)
- Adapter -- [compile one of two handler templates based on the flag](https://github.com/ap0nia/svelte.kit/blob/main/packages/adapter-aws/src/index.js#L120)

My CDK approach is different because it requires the user to invoke `cdk deploy` themselves. I was interested in building a CDK/devops centric deployment interface that didn't abstract over the underlying tools like SST, and I was referencing your project to improve it 🙏

Feel free to let me know if this was any helpful at all!

**Comment by jill64** [comment]
### Solution 1
These will be merged soon in #546

https://github.com/jill64/sveltekit-adapter-aws/blob/a0c3acf698dcd6c44793e9d65d5fb4a63de054c7/embed/external/utils/runStream.ts#L28-L32

### Solution 2
Based on @ap0nia's idea, tracked with #541

**Comment by jill64** [comment]
We implemented Solution 1 in v1.4.0 and Solution 2 in v1.5.0 and confirmed that the problem is fixed.
If you still encounter problems, please reopen.
We have also created a separate issue for other problems we found during the troubleshooting process.
